Access Node Minimum Stake Requirement (#57)
onflow/flips: protocol/20230110-accessnode-stake.md

---

# Access Node Staking Minimum

## Objective

Introduce a minimum required stake for access nodes
in the Flow protocol and staking smart contract.

## Motivation

Access nodes were proposed to be an easy way to run a Flow node
without having to commit much capital for extended periods of time.
The access node is a staked node that receives and indexes
all protocol state data and serves [the Access API.](https://developers.flow.com/nodes/access-api)

While access nodes aren't directly critical to the core functioning
of the network, they serve an important role of being the point
of access for transaction submission, state querying, and more.
They are also capable of malicious behavior that can negatively affect
users and the network. Until now, access nodes have been staked
in the identity table, but haven't been required to actually stake any FLOW
to participate.

This FLIP proposes increasing the access node minimum to a value
that reflects its importance to the network and prevents certain vulnerabilities
in the smart contract that are a result of having a zero minimum stake requirement.

The Flow community is getting ready to allow anyone to run access nodes

> without permission from the consensus committee. As part of these changes,

Reviewer: should this be?
```suggestion
without permission from the service account committee. As part of these changes,
```

the staking smart contract will be updated to remove the approval restriction
for access nodes and introduce the concept of a candidate node list,
which refers to nodes who have registered and committed sufficient stake
to be a node operator for the next epoch,
but aren't currently in the identity table.

The protocol uses this list to randomly select nodes to be node operators,
and there is a limit to the number of nodes who can be candidate nodes for each role.
Because access nodes' minimum stake requirement is currently zero,
an attacker can register any number of nodes until the limit is reached,
effectively blocking other users from registering access nodes
until the next epoch when the candidate list is reset,
at which point they could just do it again.

To attempt to address this vulnerability while also keeping access nodes relatively cheap,

> this FLIP proposes that the access node minimum stake requirement is set to .

Reviewer: ```suggestion
this FLIP proposes that the access node minimum stake requirement is set to [TBD].
```

This requirement will be treated the same as the minimum stake requirements
for other node types, meaning that access nodes will be removed if their committed
stake is less than the minimum.

**This proposal does not enable access nodes to receive rewards**.
While access nodes will have to stake FLOW, they will still not receive any rewards.

## User Benefit

> This change would not have any direct benefit to users,
> but would prevent attackers from blocking access to node registrations
> by exploiting the vulnerability.

Reviewer 2: ```suggestion
This change would benefit users operating Access Node by preventing attackers from blocking access to node registrations by exploiting the vulnerability.
```

Sounds like a benefit to me! 😀

Author: good point. 😆

## Design Proposal

> Remove any restrictions in the contract that do not count access nodes
> as staked or prevent them from staking, such as in

Reviewer: nit: access nodes currently don't commit tokens as stake, but are still "staked"
```suggestion
Remove any restrictions in the contract that do not account for access nodes
committing tokens as stake, such as in
```

`getTotalStaked`](https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowIDTableStaking.cdc#L1406)

The service account committee runs a transaction, 
[`set_minimums.cdc`](https://github.com/onflow/flow-core-contracts/blob/master/transactions/idTableStaking/admin/change_minimums.cdc)
to set the access node minimum to the desired value.

### Drawbacks

This makes running an access node slightly more expensive.

### Alternatives Considered

* Keeping the stake requirement at zero, but requiring a deposit at registration time
that is returned to the node operator at the end of the epoch.
This was discarded because it is only a temporary solution that adds

> more work but doesn't have much benefit in the short term.

Reviewer 2: ```suggestion
more engineering effort but doesn't have much benefit in the long term, because it would eventually be replaced by stake requirements similar to this proposal.
```

---

> status: Proposed 

Reviewer 3: ```suggestion
status: Accepted 
```

flip: FLIP-57
authors: Joshua Hannan (joshua.hannan@dapperlabs.com)
sponsor: Joshua Hannan (joshua.hannan@dapperlabs.com) 
updated: 2023-01-11
---

# Access Node Staking Minimum

## Objective

Introduce a minimum required stake (100 FLOW) for access nodes
in the Flow protocol and staking smart contracts.

## Motivation

Access nodes were proposed to be an easy way to run a Flow node
without having to commit much capital for extended periods of time.
The access node is a staked node that receives and indexes
all protocol state data and serves [the Access API.](https://developers.flow.com/nodes/access-api)

While access nodes aren't directly critical to the core functioning
of the network, they serve an important role of being the point
of access for transaction submission, state querying, and more.
They are also capable of malicious behavior that can negatively affect
users and the network. Until now, access nodes have been staked

> in the identity table, but haven't been required

Reviewer 4: > staked in the identity table

it might be worth-while just briefly explaining what the means

```suggestion
in the identity table (i.e. they register their node ID, address and keys with the network), but haven't been required
```

to actually stake any FLOW to participate.

This FLIP proposes increasing the access node staking minimum to a value
that reflects its importance to the network and prevents certain vulnerabilities
in the smart contract that are a result of having a zero minimum stake requirement.

The Flow community is getting ready [to allow anyone to run access nodes
without permission from the service account committee](https://github.com/onflow/flips/blob/main/protocol/20220719-automated-slot-assignment.md). As part of these changes,
the staking smart contract will be updated to remove the approval restriction
for access nodes and introduce the concept of a candidate node list,
which refers to nodes who have registered and committed sufficient stake
to be a node operator for the next epoch,
but aren't currently in the identity table.

> The protocol uses this list to randomly select nodes to be node operators,

Reviewer 4: > select nodes to be node operators

to me this sounds odd. Are we conflating nodes (machines) and node operators (humans)?

```suggestion
The protocol uses this list to randomly select nodes to participate in the network,
```

and there is a limit to the number of nodes who can be candidate nodes for each role.
Because access nodes' minimum stake requirement is currently zero,
an attacker can register any number of nodes until the limit is reached,
effectively blocking other users from registering access nodes
until the next epoch when the candidate list is reset,
at which point they could just do it again.

To attempt to address this vulnerability while also keeping access nodes relatively cheap,
this FLIP proposes that the access node minimum stake requirement is set to 100 FLOW.
This requirement will be treated the same as the minimum stake requirements
for other node types, meaning that access nodes will be removed if their committed
stake is less than the minimum.

The exact number of FLOW required for minimum stake is not that important.
It basically needs to be small enough that it is affordable
to run an access node for most people,
but not so small that it is easy to fill the candidate node list.

> The candidate node limit for access nodes will be over 1,000 nodes,

Reviewer 4: I think we are planning to change this and have a candidate list for each node role separately.

> The candidate node limit for access nodes will be over 1,000 nodes,

Author: yes, we have a candidate list for each node role

> so in order to fill up the list, it would take over 100,000 FLOW every week to perform this attack, which we believe is enough to dissuade potential attackers from exploiting it.

Reviewer 5: How can we be so convinced that the mentioned 100K FLOW is sufficient to mitigate the spam vector? If we're talking about an attack from a competing protocol, state-level actor, or otherwise highly capitalized & motivated actor, 100K seems pretty menial. How critical is the vulnerability posed by the mentioned attack vector?

A follow up question that may be out of the scope of this flip: are there fallback mechanisms the community can leverage to recover from this attack should it occur? Such a mechanism could minimize the risk posed by setting an accessible minimum which I realize is important for community access & participation.

Reviewer 6: Hey @sisyphusSmiling, like @vishalchangrani explained, the vulnerability posed is not huge and only exists at the networking layer. However, as the number of slots in the registration list (FlowIDTable) is limited, to ensure that attackers don’t acquire all slots and bar anyone else from being considered to run a node, we decided to introduce a positive yet nominal minimum stake (100 Flow) per slot for an operator to legitimately acquire a slot in the FlowIDTable. After all interested operators have registered on the FlowIDTable and have submitted at least the minimum required stakes to the addNodeInfo function, an algorithm randomly selects some operators from the filled slots, allowing them to run an access node in the subsequent epoch.

Reviewer 5: Thank you both @vishalchangrani and @KshitijChaudharyHBS for the explanation. The mechanisms you noted as well as the [stated plans](https://github.com/onflow/flips/pull/57#issuecomment-1381995597) for upcoming staking auctions addressed any concerns I had.

Reviewer 7: > While it cannot impact safety, an errant Access node can impact liveness 

@vishalchangrani what does this mean exactly ?

Reviewer 3: hi @bluesign - I mean that errant access nodes can't cause a fork or affect the validity of the chain however they could cause the chain to slow down or even stop block production.

(Specifically, “safety” can be thought of as nothing incorrect ever happening; and “liveness” can be thought of as the correct thing always eventually happening [source](https://learn.radixdlt.com/article/what-do-security-safety-and-liveness-mean#:~:text=For%20blockchains%20and%20DLTs%2C%20these,correct%20thing%20always%20eventually%20happening.)

**This proposal does not enable access nodes to receive rewards**.
While access nodes will have to stake FLOW, they will still not receive any rewards.

## User Benefit

This change would benefit users operating Access Node by preventing attackers from blocking access to node registrations by exploiting the vulnerability.

## Design Proposal

Remove any restrictions in the contract that do not account for access nodes committing tokens as stake, such as in

> `getTotalStaked`](https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowIDTableStaking.cdc#L1406)

Reviewer 4: ```suggestion
[`getTotalStaked`](https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowIDTableStaking.cdc#L1406)
```

## Objective

Introduce a minimum required stake (100 FLOW) for access nodes
in the Flow protocol and staking smart contracts.

## Motivation

Access nodes were proposed to be an easy way to run a Flow node
without having to commit much capital for extended periods of time.
The access node is a staked node that receives and indexes
all protocol state data and serves [the Access API.](https://developers.flow.com/nodes/access-api)

While access nodes aren't directly critical to the core functioning
of the network, they serve an important role of being the point
of access for transaction submission, state querying, and more.
They are also capable of malicious behavior that can negatively affect
users and the network. Until now, access nodes have been staked
in the identity table, but haven't been required
to actually stake any FLOW to participate.

This FLIP proposes increasing the access node staking minimum to a value
that reflects its importance to the network and prevents certain vulnerabilities
in the smart contract that are a result of having a zero minimum stake requirement.

The Flow community is getting ready [to allow anyone to run access nodes

> without permission from the consensus committee](https://github.com/onflow/flips/blob/main/protocol/20220719-automated-slot-assignment.md). As part of these changes,

Reviewer: should this be?
```suggestion
without permission from the service account committee](https://github.com/onflow/flips/blob/main/protocol/20220719-automated-slot-assignment.md). As part of these changes,
```
